### PR TITLE
Implement LocalMinistep as C++

### PR DESCRIPTION
**Pull request by joakim-hove**
Implement `LocalMinistep`as proper C++ and wrap with Pybind.

On top of: https://github.com/equinor/ert/pull/2966

### CODE REVIEW

File: libres/lib/enkf/enkf_obs.cpp [hunk 1]
```diff
@@ -440,15 +440,15 @@ static void enkf_obs_get_obs_and_measure_node(
...
                                        const int_vector_type *ens_active_list,
                                        meas_data_type *meas_data,
                                        obs_data_type *obs_data) {
 
     int iobs;
-    for (iobs = 0; iobs < local_obsdata_get_size(local_obsdata); iobs++) {
-        const auto *obs_node = local_obsdata_iget(local_obsdata, iobs);
-        enkf_obs_get_obs_and_measure_node(enkf_obs, fs, obs_node,
+    for (iobs = 0; iobs < local_obsdata->size(); iobs++) {
+        const auto &obs_node = local_obsdata->operator[](iobs);
```
Comment (pinkwah) [inline libres/lib/enkf/enkf_obs.cpp:450]:
```suggestion
        const auto &obs_node = (*local_obsdata)[iobs];
```

File: libres/lib/enkf/local_config.cpp [hunk 2]
```diff
@@ -247,3 +229,33 @@ local_config_get_updatestep(const local_config_type *local_config) {
 
     return updatestep;
 }
+
+namespace {
+
+LocalObsData &get_obsdata(py::handle obj, const std::string &obs_key) {
+    auto *local_config = reinterpret_cast<local_config_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (pinkwah) [inline libres/lib/enkf/local_config.cpp:237]:
```suggestion
    auto *local_config = ert::from_cwrap<local_config_type>(obj);
```

File: libres/lib/enkf/local_config.cpp [hunk 3]
```diff
@@ -247,3 +229,33 @@ local_config_get_updatestep(const local_config_type *local_config) {
...
+    auto *obsdata = local_config_get_obsdata(local_config, obs_key.c_str());
+    return *obsdata;
+}
+
+LocalObsData copy_obsdata(py::handle obj, const std::string &obs_key,
+                          const std::string &target_key) {
+    auto *local_config = reinterpret_cast<local_config_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (pinkwah) [inline libres/lib/enkf/local_config.cpp:245]:
```suggestion
    auto *local_config = ert::from_cwrap<local_config_type>(obj);
```

File: libres/lib/enkf/local_config.cpp [hunk 4]
```diff
@@ -247,3 +229,33 @@ local_config_get_updatestep(const local_config_type *local_config) {
...
+}
+
+} // namespace
+
+RES_LIB_SUBMODULE("local.local_config", m) {
+    using namespace py::literals;
+    m.def("get_obsdata_ref", &get_obsdata, "self"_a, "key"_a,
+          py::return_value_policy::reference);
+    m.def("get_obsdata_copy", &copy_obsdata, py::arg("self"), py::arg("key1"),
+          py::arg("key2"));
```
Comment (pinkwah) [inline libres/lib/enkf/local_config.cpp:260]:
```suggestion
    m.def("get_obsdata_copy", &copy_obsdata, "self"_a, "key1"_a,
          "key2"_a);
```

File: libres/lib/enkf/local_ministep.cpp [hunk 5]
```diff
@@ -64,14 +64,14 @@ void local_ministep_free__(void *arg) {
...
+                                LocalObsData *obsdata) {
     if (ministep->observations == NULL)
         ministep->observations = obsdata;
     else { // Add nodes from input observations to existing observations
         int iobs;
-        for (iobs = 0; iobs < local_obsdata_get_size(obsdata); iobs++) {
-            auto *obs_node = local_obsdata_iget(obsdata, iobs);
-            LocalObsDataNode new_node(*obs_node);
+        for (iobs = 0; iobs < obsdata->size(); iobs++) {
+            auto &obs_node = obsdata->operator[](iobs);
```
Comment (pinkwah) [inline libres/lib/enkf/local_ministep.cpp:73]:
```suggestion
            auto &obs_node = (*obsdata)[iobs];
```

File: libres/lib/enkf/local_ministep.cpp [hunk 6]
```diff
@@ -156,6 +155,19 @@ void add_obsdata_node(py::handle self, LocalObsDataNode &node) {
     local_ministep_add_obsdata_node(ministep, &node);
 }
 
+LocalObsData &get_obsdata(py::handle obj) {
+    auto *ministep = reinterpret_cast<local_ministep_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (pinkwah) [inline libres/lib/enkf/local_ministep.cpp:160]:
```suggestion
    auto *ministep = ert::from_cwrap<local_ministep_type>(obj);
```

File: libres/lib/enkf/local_ministep.cpp [hunk 7]
```diff
@@ -156,6 +155,19 @@ void add_obsdata_node(py::handle self, LocalObsDataNode &node) {
...
+    auto *obsdata_ptr = local_ministep_get_obsdata(ministep);
+    return *obsdata_ptr;
+}
+
+void attach_obsdata(py::handle obj, LocalObsData &obsdata) {
+    auto *ministep = reinterpret_cast<local_ministep_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (pinkwah) [inline libres/lib/enkf/local_ministep.cpp:167]:
```suggestion
    auto *ministep = ert::from_cwrap<local_ministep_type>(obj);
```

File: libres/lib/enkf/obs_data.cpp [hunk 8]
```diff
@@ -334,6 +346,18 @@ obs_data_type *obs_data_alloc(double global_std_scaling) {
     return obs_data;
 }
 
+obs_data_type *obs_data_alloc_copy(const obs_data_type *src) {
```
Comment (pinkwah) [inline libres/lib/enkf/obs_data.cpp:349]: This function does not seem to be in use. Delete it?

File: libres/lib/include/ert/enkf/local_ministep.hpp [hunk 9]
```diff
@@ -42,7 +42,7 @@ class local_ministep_type {
     UTIL_TYPE_ID_DECLARATION;
     std::string
         name; /* A name used for this ministep - string is also used as key in a hash table holding this instance. */
-    local_obsdata_type *observations;
+    LocalObsData *observations;
```
Comment (pinkwah) [inline libres/lib/include/ert/enkf/local_ministep.hpp:45]: Consider making this a value or use `std::shared_ptr`.

File: res/enkf/enkf_main.py [hunk 10]
```diff
@@ -340,6 +340,7 @@ def getModelConfig(self) -> ModelConfig:
     def getLocalConfig(self) -> LocalConfig:
         """@rtype: LocalConfig"""
         config = self._get_local_config().setParent(self)
+        print("Have fetched local config")
```
Comment (pinkwah) [inline res/enkf/enkf_main.py:343]: Remove

File: res/enkf/local_obsdata.py [hunk 11]
```diff
@@ -1,138 +1,3 @@
-from cwrap import BaseCClass
+from res._lib.local.local_obsdata import LocalObsdata
```
Comment (pinkwah) [inline res/enkf/local_obsdata.py:1]: I think you can remove this file entirely.

File: tests/libres_tests/res/enkf/test_local_config.py [hunk 12]
```diff
@@ -104,14 +104,14 @@ def test_all_active(self):
             main = test_context.getErt()
 
             local_config = main.getLocalConfig()
-
+            print("Have returned local config")
```
Comment (pinkwah) [inline tests/libres_tests/res/enkf/test_local_config.py:107]: Debug stuff

File: tests/libres_tests/res/enkf/test_local_config.py [hunk 13]
```diff
@@ -75,7 +75,7 @@ def test_local_obs_data(self):
             self.assertTrue(node_added)
 
             # Error when adding existing obs node
-            with self.assertRaises(KeyError):
+            with self.assertRaises(Exception):
```
Comment (pinkwah) [inline tests/libres_tests/res/enkf/test_local_config.py:78]: You can use `throw py::key_error` in pybind11 C++ code to get a `KeyError` here

**Review by pinkwah — COMMENTED** [review]
I'm not sure if this PR is ready for review, but here are some minor comments

File: tests/libres_tests/res/enkf/test_local_config.py (hunk 13 above)
Comment (joakim-hove) [reply tests/libres_tests/res/enkf/test_local_config.py:78]: That I did not know - that was very valuable information.

File: libres/lib/enkf/enkf_obs.cpp (hunk 1 above)
Comment (joakim-hove) [reply libres/lib/enkf/enkf_obs.cpp:450]: OK - in #2966

File: libres/lib/enkf/local_config.cpp (hunk 2 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_config.cpp:237]: Fixed in #2966.

File: libres/lib/enkf/local_config.cpp (hunk 4 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_config.cpp:260]: Fixed in https://github.com/equinor/ert/pull/2966.

File: libres/lib/enkf/local_config.cpp (hunk 3 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_config.cpp:245]: Fixed in https://github.com/equinor/ert/pull/2966.

File: libres/lib/enkf/local_ministep.cpp (hunk 6 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_ministep.cpp:160]: Fixed in https://github.com/equinor/ert/pull/2966.

File: libres/lib/enkf/obs_data.cpp (hunk 8 above)
Comment (joakim-hove) [reply libres/lib/enkf/obs_data.cpp:349]: It is in use in #2981 - although maybe not in the sorry state of this actual PR.

**Comment by joakim-hove** [comment]
I'll close this now - after multiple  rounds of rebasing this PR has been completely molested. The next step which should be reviewed and merged is #2966. When that is merged the final step is #2981.

**Pull request closed without merging**